src/test/java/com/microsoft/sqlserver/jdbc/fedauth/ConcurrentLoginTest.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Microsoft JDBC Driver for SQL Server Copyright(c) Microsoft Corporation All rights reserved. This program is made
+ * available under the terms of the MIT License. See the LICENSE file in the project root for more information.
+ */
+package com.microsoft.sqlserver.jdbc.fedauth;
+
+import static org.junit.jupiter.api.Assertions.fail;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+import org.junit.platform.runner.JUnitPlatform;
+import org.junit.runner.RunWith;
+
+import com.microsoft.sqlserver.jdbc.SQLServerDataSource;
+import com.microsoft.sqlserver.testframework.Constants;
+
+
+@RunWith(JUnitPlatform.class)
+@Tag(Constants.Fedauth)
+public class ConcurrentLoginTest extends FedauthCommon {
+
+    @Test
+    public void testConcurrentLogin() {
+        Random rand = new Random();
+        int numberOfThreadsForEachType = rand.nextInt(15) + 1; // 1 to 15
+
+        for (int i = 0; i < numberOfThreadsForEachType; i++) {
+            // Access token based authentication
+            new Thread() {
+                public void run() {
+                    try {
+                        SQLServerDataSource ds = new SQLServerDataSource();
+                        ds.setServerName(azureServer);
+                        ds.setDatabaseName(azureDatabase);
+                        ds.setAccessToken(accessToken);
+
+                        try (Connection conn = ds.getConnection()) {
+                            testUserName(conn, azureUserName, SqlAuthentication.NotSpecified);
+                        }
+                    } catch (SQLException e) {
+                        fail(e.getMessage());
+                    }
+                }
+            }.start();

```

Comment:
This should probably use an executor instead of multiple thread, but assuming we're going to proceed with Threads, we can clean up the syntax.

```java
Runnable r1 = () -> {
    try {
        SQLServerDataSource ds = new SQLServerDataSource();
        ds.setServerName(azureServer);
        ds.setDatabaseName(azureDatabase);
        ds.setAccessToken(accessToken);

        try (Connection conn = ds.getConnection()) {
            testUserName(conn, azureUserName, SqlAuthentication.NotSpecified);
         }
    } catch (SQLException e) {
          fail(e.getMessage());
    }
}
// define r2, r3

(for int i = 0; i < numberOfThreadsForEachType; i++) {
    new Thread(r1).start();
    new Thread(r2).start();
    if (isWindows && enableADIntegrated)
        new Thread(r3).start();
}
```